beacon/light/request/scheduler.go

```diff
@@ -0,0 +1,371 @@
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package request
+
+import (
+	"math"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common/mclock"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// Module represents a mechanism which is typically responsible for downloading
+// and updating a passive data structure.
+// Modules can start network requests through Tracker and receive request events
+// related to the sent requests that can signal a response, a failure or a timeout.
+// They also receive server-related events. Note that they do not directly interact
+// with servers but may keep track of certain parameters of registered servers,
+// based on the received server events. These server parameters may affect the
+// possible range of requests to be sent to a given server.
+// Modules are called by Scheduler whenever a global trigger is fired. All events
+// fire the trigger. Modules themselves can also self-trigger, ensuring an
+// immediate next processing round after the target data structure has been
+// changed in a way that could make further actions possible either by the same
+// or another Module.
+type Module interface {
+	// Process is a non-blocking function that is called on each Module whenever
+	// a processing round is triggered. It can start new requests through the
+	// received Tracker, process events and/or do other data processing tasks.
+	// Note that request events are only passed to the module that made the given
+	// request while server events are passed to every module.
+	//
+	// Note: Process functions of different modules are never called concurrently;
+	// they are called by Scheduler in the same order of priority as they were
+	// registered in.
+	Process([]Event)
+	MakeRequest(Server) (Request, float32)
+}
+
+// Scheduler is a modular network data retrieval framework that coordinates multiple
+// servers and retrieval mechanisms (modules). It implements a trigger mechanism
+// that calls the Process function of registered modules whenever either the state
+// of existing data structures or events coming from registered servers could
+// allow new operations.
+type Scheduler struct {
+	lock    sync.Mutex
+	clock   mclock.Clock
+	modules []Module // first has highest priority
+	names   map[Module]string
+	servers map[server]struct{}
+	targets map[targetData]uint64
+
+	pending       map[ServerAndID]pendingRequest
+	eventLock     sync.Mutex
```

Comment:
Done, I also made this part a lot nicer because it was indeed a bit messy. Now the lock is only used by `addEvent` and `filterEvents`, ensuring deadlock-free server event callbacks.